src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java.g

```diff
@@ -417,12 +421,24 @@ builtInTypeSpec[boolean addImagNode]
         }
     ;
 
+typeSpecBracketsAndAnnotationsNonGreedy
+    :
+     (
+        ({LA(1) == AT}? annotations
+        | )
+         brackets )*
+    ;
+
 typeSpecBracketsAndAnnotations
     :
      (options{greedy=true; }:
         ({LA(1) == AT}? annotations
         | )
-            (lb:LBRACK^ {#lb.setType(ARRAY_DECLARATOR);} rb:RBRACK)  )*
+         brackets  )*
```

Comment:
Using the `brackets` rule allows us to keep `[]` brackets as a parallel element in relation to each other here, and in all productions where the new `brackets` rule is used. This corresponds to java grammar:

```
ArrayCreatorRest:
    [ (] {[]} ArrayInitializer  |  Expression ] {[ Expression ]} {[]})

```